Sources/ApplicativeRouter/PartialIso.swift

```diff
@@ -0,0 +1,98 @@
+import Prelude
+
+// todo: use a profunctor Iso?
+//typealias Iso_<S, T, A, B> = ((S) -> A) -> ((B) -> T)
+
+// todo: move to prelude?
+
+// TODO: Change PartialIso to be Partial<A, B, E: Semigroup>: `((A) -> Validation<E, B>, (B) -> Validation<E, A>)
+
+/// A partial isomorphism is a partial function that is invertible on the parts it is defined, i.e.
+/// `image(x) = y` if and only `preimage(y) = x`.
+public struct PartialIso<A, B> {
```

Comment:
this is the core of what `PartialIso` is, and feel we should move to its own package in Prelude